Makefile

```diff
@@ -1,53 +1,75 @@
-SHELL = /bin/sh
-GO = go
-PROTOC = protoc
-CAPNPC = capnpc
+SHELL := /bin/sh
 
-HERE = $(shell pwd)
-BIN = $(HERE)/bin
-GPM = $(HERE)/gpm
-DEPS = $(HERE)/.godeps
-GOPATH = $(DEPS):$(HERE)
-TOOLS = $(shell GOPATH=$(GOPATH) $(GO) tool)
+PROTOC := protoc
+CAPNPC := capnpc
 
-PLATFORM=$(shell uname)
+BIN := $(CURDIR)/bin
+DEPS := $(CURDIR)/.godeps
 
 # Setup commands and env vars if there is no system go linked into bin/go
-PATH := $(HERE)/bin:$(DEPS)/bin:$(PATH)
-
-PACKAGE = github.com/mozilla-services/pushgo
-PREFIX = src/$(PACKAGE)/simplepush
-TARGET = simplepush
-COVER_MODE = count
-COVER_PATH = $(HERE)/.coverage
-
-VERSION = $(strip $(shell [ -e $(HERE)/GITREF ] && cat $(HERE)/GITREF 2>/dev/null))
-ifeq ($(VERSION),)
-	VERSION = $(strip $(shell git describe --tags --always HEAD 2>/dev/null))
-endif
-
-ifneq ($(strip $(VERSION)),)
-	GOLDFLAGS := -X $(PACKAGE)/simplepush.VERSION $(VERSION) $(GOLDFLAGS)
-endif
-
-
-.PHONY: all build clean test $(TARGET) memcached
+PATH := $(CURDIR)/bin:$(DEPS)/bin:$(PATH)
+GOPATH := $(DEPS):$(CURDIR)
+GPM := GOPATH=$(GOPATH) $(CURDIR)/gpm
+GO := GOPATH=$(GOPATH) go
+
+PLATFORM := $(shell uname)
+
+# All server packages.
+PACKAGE := github.com/mozilla-services/pushgo
+SUBPACKAGES := $(addprefix $(PACKAGE)/,id retry simplepush)
+
+# The executable name.
+TARGET := simplepush
+
+# Generated Protobuf and Cap'n Proto targets.
+GEN_TARGETS := log_message.pb.go routable.capnp.go
+GEN_PATHS := $(addprefix $(CURDIR)/src/$(PACKAGE)/simplepush/,\
+	$(GEN_TARGETS))
+
+# Interfaces for mocking.
+INTERFACES := config.go worker.go storage.go locator.go metrics.go\
+	balancer.go socket.go handlers.go log.go router.go proprietary_ping.go
+MOCKS := $(addprefix src/$(PACKAGE)/simplepush/,\
+	$(patsubst %.go,mock_%_test.go,$(INTERFACES)))
+
+# Coverage mode; can be "set", "count", or "atomic". See
+# https://blog.golang.org/cover for more info.
+COVER_MODE := count
+
+# Temporary coverage profile paths for each package, and the merged profile
+# targets.
+COVER_PATH := $(CURDIR)/.coverage
+COVER_PACKAGES := $(addprefix $(COVER_PATH)/,\
+	$(addsuffix %.out,$(SUBPACKAGES)))
+COVER_TARGETS := coverage.out coverage.server.out
+COVER_HTML_TARGETS := $(patsubst %.out,%.html,$(COVER_TARGETS))
+
+.PHONY: all build gen clean-gen $(TARGET) test-mocks clean-mocks test \
+	test-server test-gomc test-gomemcache check-cov travis-cov\
+	html-cov html-server-cov clean-cov bench bench-server vet clean
+.INTERMEDIATE: $(COVER_TARGETS)
 
 all: build
 
 $(BIN):
 	mkdir -p $(BIN)
 
+# Fetch and install dependencies.
 $(DEPS):
 	@echo "Installing dependencies"
-	GOPATH=$(GOPATH) $(GPM) install
-	GOPATH=$(GOPATH) $(GO) install github.com/gogo/protobuf/... \
+	$(GPM) install
+	$(GO) install github.com/gogo/protobuf/... \
```

Comment:
hrm. gpm should be able to install everything (you don't include the version if you don't want to vendor lock the library). Why are we calling gpm, then installing these libraries again?